src/platform/GenericOTARequestorDriver.cpp

```diff
@@ -46,29 +64,93 @@ uint16_t GenericOTARequestorDriver::GetMaxDownloadBlockSize()
+{
+    if ((a.fabricIndex == b.fabricIndex) && (a.providerNodeID == b.providerNodeID) && (a.endpoint == b.endpoint))
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+void GenericOTARequestorDriver::HandleError(UpdateFailureState state, CHIP_ERROR error) {}
+
+void GenericOTARequestorDriver::HandleIdleState()
 {
-    // TODO: Schedule the next QueryImage
+    // Default provider timer runs if and only if the OTARequestor's update state is kIdle.
+    // Must (re)start the timer every time we enter the kIdle state
+    StartDefaultProviderTimer();
 }
 
 void GenericOTARequestorDriver::UpdateAvailable(const UpdateDescription & update, System::Clock::Seconds32 delay)
 {
+    // IMPLEMENTATION CHOICE:
+    // This implementation unconditionally downloads an available update
+
     VerifyOrDie(mRequestor != nullptr);
-    ScheduleDelayedAction(UpdateFailureState::kDownloading, delay,
-                          [](System::Layer *, void * context) { ToDriver(context)->mRequestor->DownloadUpdate(); });
+    ScheduleDelayedAction(
+        delay, [](System::Layer *, void * context) { ToDriver(context)->mRequestor->DownloadUpdate(); }, this);
 }
 
 void GenericOTARequestorDriver::UpdateNotFound(UpdateNotFoundReason reason, System::Clock::Seconds32 delay)
 {
     VerifyOrDie(mRequestor != nullptr);
 
+    ProviderLocation::Type providerLocation;
+    bool willTryAnotherQuery = false;
+
+    switch (reason)
+    {
+    case UpdateNotFoundReason::UpToDate:
+        willTryAnotherQuery = false;
+        break;
+
+    case UpdateNotFoundReason::Busy:
+        willTryAnotherQuery = true;
+        break;
+
+    case UpdateNotFoundReason::ConnectionFailed:
+    case UpdateNotFoundReason::NotAvailable:
+        // IMPLEMENTATION CHOICE:
+        // This implementation schedules a query only if a different provider is available
+        if ((DetermineProviderLocation(providerLocation) != true) || ProviderLocationsEqual(providerLocation, mLastUsedProvider))
+        {
+            willTryAnotherQuery = false;
+        }
+        else
+        {
+            willTryAnotherQuery = true;
+        }
+        mRequestor->SetCurrentProviderLocation(providerLocation);
+        mLastUsedProvider = providerLocation;
+        break;
+
+    default:
+        willTryAnotherQuery = false;
+        break;
+    }
+
     if (delay < kDefaultDelayedActionTime)
     {
         delay = kDefaultDelayedActionTime;
     }
-
-    ScheduleDelayedAction(UpdateFailureState::kQuerying, delay,
-                          [](System::Layer *, void * context) { ToDriver(context)->mRequestor->TriggerImmediateQuery(); });
+    if (willTryAnotherQuery == true)
+    {
+        ChipLogProgress(SoftwareUpdate, "UpdateNotFound, scheduling a retry");
+        ScheduleDelayedAction(delay, StartDelayTimerHandler, this);
+    }
+    else
+    {
+        ChipLogProgress(SoftwareUpdate, "UpdateNotFound, not scheduling further retries");
```

Comment:
We should call `ClearCurrentProviderLocation` here

Reply:
Done

Reply:
Fixed